rp2040-hal/examples/dma.rs

```diff
@@ -0,0 +1,91 @@
+//! # Memory to memory DMA transfer Example
+//!
+//! This application demonstrates how to use DMA to transfer data from memory to memory buffers.
+//!
+//! See the `Cargo.toml` file for Copyright and licence details.
+#![no_std]
+#![no_main]
+
+use cortex_m::singleton;
+use cortex_m_rt::entry;
+use embedded_hal::digital::v2::OutputPin;
+use hal::dma::{single_buffer, DMAExt};
+use hal::pac;
+use panic_halt as _;
+use rp2040_hal as hal;
+use rp2040_hal::clocks::Clock;
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+#[entry]
+fn main() -> ! {
+    let mut pac = pac::Peripherals::take().unwrap();
+    let core = pac::CorePeripherals::take().unwrap();
+
+    // Setup clocks and the watchdog.
+    let mut watchdog = hal::watchdog::Watchdog::new(pac.WATCHDOG);
+    let clocks = hal::clocks::init_clocks_and_plls(
+        XTAL_FREQ_HZ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    // Setup the pins.
+    let sio = hal::sio::Sio::new(pac.SIO);
+    let pins = hal::gpio::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // Initialize DMA.
+    let dma = pac.DMA.split(&mut pac.RESETS);
+    // Configure GPIO25 as an output
+    let mut led_pin = pins.gpio25.into_push_pull_output();
+    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());
+
+    // Use DMA to transfer some bytes (single buffering).
+    let tx_buf = singleton!(: [u8; 16] = [0x42; 16]).unwrap();
+    let rx_buf = singleton!(: [u8; 16] = [0; 16]).unwrap();
+
+    // Use a single_buffer to read from tx_buf and write into rx_buf
+    let transfer = single_buffer::Config::new(dma.ch0, tx_buf, rx_buf).start();
+    // Wait for both DMA channels to finish
+    let (_ch, tx_buf, rx_buf) = transfer.wait();
+
+    // Compare buffers to see if the data was transferred correctly
+    for i in 0..tx_buf.len() {
+        if rx_buf[i] != tx_buf[i] {
+            // Fast blink on error
+            loop {
+                led_pin.set_high().unwrap();
+                delay.delay_ms(100);
+                led_pin.set_low().unwrap();
+                delay.delay_ms(100);
+            }
+        }
+    }
+
+    // Slow blink on success
+    loop {
+        led_pin.set_high().unwrap();
+        delay.delay_ms(500);
+        led_pin.set_low().unwrap();
+        delay.delay_ms(500);
+    }
```

Comment:
You can compare arrays directly like this:

```suggestion
    let delay = if tx_buf == rx_buf {
        500
    } else {
        100
    };
    
    // Slow blink on success
    loop {
        led_pin.set_high().unwrap();
        delay.delay_ms(delay);
        led_pin.set_low().unwrap();
        delay.delay_ms(delay);
    }
```

See [here](https://godbolt.org/z/93zjss7W4) for the output.

Reply:
I thought I'd check the asm output of this vs the iter().zip way I wrote the on-target-test version on this.
[Lowering to memcmp feels very cheaty](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATBtKOAMngMmABy7gBGmMQgAKykAA6oCoR2DC5uHt6%2Bicm2AoHBYSyR0XFWmDapQgRMxATp7p4%2BlpjWeQzVtQQFoRFRsZY1dQ2ZzeZdPUUlsQCUlqgmxMjsHACkXgDMYGCrMQBCDKgA%2BubouwAiqxoAglfX8SbhANRUDC%2BoqBAzTwC0qxvYJ7hD60J6rADsezuTxhT3oBCeTDBG3OYP2ki8JgAHP89lwAGwXXHQ2HwoHI1G7PYaXEEokbKE3WFPNAseK1TAQdb4pikbnhGZ3CGXG5CzZU5BUYAQWrEJgAT0wAEdBTERfdHi83qz2cROUwQGCvIT9tjaSbzqRwobuVSzQy6Wrvn8AUCQRDGddmUj/pcUUCheD1WKNhKpTLiHL5XgsCqiTcHs9XizUGyOTKbca7TiHRarZmTXt7XiLc7/oDga4PSSYfxiE88A23hoAHQtpgt%2BgML5gyE15kNqiIql4C5Pba%2BgP7Udq3uegcLp56giLN5UMRKYlMxfC/u77cwgimTCB4Oi8X7SXS2UKwhRZWq9WJrUptN6jNGwvFx2W62f7PmhcZaupWtDVge7z1hAqikE88rfEEiItnexBfC2ABeeDxBA4TIQQURfN84FeoueBDqo447P68pzv2i5LpgK7EGuG4ngydGwvuJEwlxzL9keJgnqKQYcHMtCcDEvCeBwWikKgnAAEpmAiCgLEsmBGhsPCkAQmiiXMADW3gxC2MRcOCWkxPiGz4hoZn4l44L6JwkhSXpcmcLwCggL4ukyaJpBwLASA6nQUTkJQoX0NExAEuCvhYAAbngywAGp4JgADuADy8SMJw2k0LQ%2BHEN5OHueEQS1PKBW8JVzDEPK2XhNoFR%2BdprJsII2UMLQNX%2BaQWAsIYwDiAN%2BB6pUiWYN5A2YKoFQmPhtXkIIrTubQeDhHKjUuFg7lHngLC1XMVAGMACjpVluX5dwvD8IIIhiOwUgyIIigqOoA26Fw%2BgjSApjmPoW3eZAcyoPE7SzT8ThPBDBA/PQ01gSiCi/D8QzAIxvoEAgESJfiLA/IcwQ/JgTCVejV5Gl4N6Kkq6xeF5rRtakDgMM4riNHoARBL0xT9L9OQpAIIyeELSQiwwkx9NEv3lJUAidMMXOZPLLOKx0QzdHzUyC4MXRi3o4x1DLAty3MqmLMsEhiRJbkDfJHBLspyBPFwLb4uCLYaE8EC4IQJCaVwMy8H5WgzHMCDk1g0RfKQRleCZGxeFiWIpwAnDEkh2eCGfgvizkcK5pDHVwGi%2BNJslO15Pk6XpcxBYgKCpvEYVkBQEBRf0sVewlmDJWlGU5Xl0mFXQJVldaA31dVK2z41zWtTYK2dYwBA9X17lDSNY2yRNrPTbNsnzYty13at%2BHiQNm3bdVe0rLJh3HXdp3nZdw83WP92yE94ivQ9eQSg1DuV0BsP6RgAbKWBuEUG8d4apGhrDeGiMB6tF9GjH4GNahYwIDjPGxQCZExJpgMmFNGxYOpozOmMZlSM2Zm0NmEBHBG1%2BrzQoss9DC3aKwhIkt2hm2mOrRhStta8IVu0ZWOsOHm2NmI1W4sDam11pwkO8xrYvTthwSSpAq68Cdi7cwbsPZex9n7AORB6zrC0qHeu/lI4J28JIFsXAYgbCxGZHOGcsQZw0BsDO4Dr4lzLhXXR7ka6WDruHAKTcIAhVbu3CKXcEnRQBnFfug9MBXRHrdcexUohTwqlVRq89ilNRam1VeqYuob16v1femBhpGD3rwA%2BU0ZruVPsgJaKxtJBCvhtLaO15QPwOsQI6J0%2BDv2yV/FagC/4vWkIAj6IDvreAgcYaBt84Hg0hogzgMM4aQ1QcjDB6NMbYxRLjfGhNiYCFIeTSmlCpQ0xoSVOhXgmYtBEZ4ZhHNeHsP5kIvhuRUi8O4akQR%2BsJFVHkRkRRMLRETBUbI%2BWcLuZouRTIoRls1I2zUdfHReiPLO0BgQYxntva%2B39vgSxwdbHRIcYnOIQSHbV08pE3yDctFfOJRExlcxpqlTZpIIAA%3D) but I like it.